### DISCUSSION TITLE
Push nuget package to Github.com: The response ended prematurely.An error occurred while sending the request

### DISCUSSION THREAD

**Discussion by PieterjanDeClippel** [community/community#24915] [Packages]
<p>I’m trying to push NuGet packages to <a href="http://Github.com" rel="nofollow noopener">Github.com</a>. I’ve successfully configured an action to build, test and push a .NET Standard package:</p>
<pre><code class="lang-auto">name: .NET Core

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: windows-latest

    steps:
    - name: Checkout
      uses: actions/checkout@v2

    - name: Setup .NET Core
      uses: actions/setup-dotnet@v1.5.0
      with:
        dotnet-version: 3.1.301
        # Authenticates packages to push to GPR
        source-url: https://nuget.pkg.github.com/${{ github.repository_owner }}/index.json
      env:
        NUGET_AUTH_TOKEN: '%NUGET_AUTH_TOKEN%'

    - name: Install dependencies
      run: dotnet restore

    - name: Build
      run: dotnet build --configuration Release --no-restore

    - name: Test
      run: dotnet test --no-restore --verbosity normal

    - name: Pack
      run: dotnet pack --no-build --configuration Release MintPlayer.SeasonChecker/MintPlayer.SeasonChecker.csproj --output .

    - name: PushNuget
      run: dotnet nuget push *.nupkg --source https://api.nuget.org/v3/index.json --api-key ${{ secrets.PUBLISH_TO_NUGET_ORG }} --skip-duplicate

    - name: PushGithub
      run: dotnet nuget push *.nupkg --no-symbols --skip-duplicate
      env:
        NUGET_AUTH_TOKEN: ${{ github.token }}
</code></pre>
<p>The package I’m trying to push now, is a multi-target nuget package, but that shouldn’t really matter. My workflow for this other package is the following:</p>
<pre><code class="lang-auto">name: .NET Core

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:

    runs-on: windows-latest

    steps:
    - name: Checkout
      uses: actions/checkout@v2
    
    - name: Setup .NET Core
      uses: actions/setup-dotnet@v1.5.0
      with:
        dotnet-version: 3.1.301
        # Authenticates packages to push to GPR
        source-url: https://nuget.pkg.github.com/${{ github.repository_owner }}/index.json
      env:
        NUGET_AUTH_TOKEN: '%NUGET_AUTH_TOKEN%'
    
    - name: Setup MSBuild
      uses: microsoft/setup-msbuild@v1.0.1
    
    - name: Install dependencies
      run: msbuild /t:Restore
      env:
        NUGET_AUTH_TOKEN: ${{ github.token }}
    
    - name: Build
      run: msbuild /t:Pack /p:Configuration=Debug Library/MintPlayer.MVVM/MintPlayer.MVVM.csproj
    
    - name: Copy
      run: copy Library/MintPlayer.MVVM/bin/Debug/*.nupkg .
    
    - name: PushNuget
      run: dotnet nuget push *.nupkg --source https://api.nuget.org/v3/index.json --api-key ${{ secrets.PUBLISH_TO_NUGET_ORG }} --skip-duplicate
    - name: PushGithub
      run: dotnet nuget push *.nupkg --no-symbols --skip-duplicate
      env:
        NUGET_AUTH_TOKEN: ${{ github.token }}
</code></pre>
<p>So to build the package, I’m using MSBuild, but to push the package I’m still using the ordinary dotnet commands. The push to <a href="http://nuget.org" rel="nofollow noopener">nuget.org</a> works fine, but the push to <a href="http://Github.com" rel="nofollow noopener">Github.com</a> fails with the following error:</p>
<pre><code class="lang-auto">An error was encountered when fetching 'PUT https://nuget.pkg.github.com/MintPlayer/'. The request will now be retried.
An error occurred while sending the request.
The response ended prematurely.
PUT https://nuget.pkg.github.com/MintPlayer/
</code></pre>
<p>Why is this push to my GitHub feed failing?</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/da39bd32ee01e45136a4446f8ae068311a1f6952.png" data-download-href="/uploads/short-url/v8vNdPlNCF1hjsIbQShri0ygL86.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/9629574/181087693-5f7ae2be-4d8f-48df-8311-9b4a2dba0759.png" alt="image" data-base62-sha1="v8vNdPlNCF1hjsIbQShri0ygL86" width="690" height="485" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/da39bd32ee01e45136a4446f8ae068311a1f6952_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">780×549 20.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by PieterjanDeClippel** [comment]
<p>I’ve narrowed down the issue.</p>
<p>I’m testing with the same repository, same code, same state. On each test, I’m incrementing the Nuget version of my package. Sometimes it succeeds in pushing to GPR, but most of the time it fails to push the new package.</p>
<p>Pushing to <a href="http://nuget.org" rel="nofollow noopener">nuget.org</a> has never been a problem in my tests, Pushing to <a href="http://github.com" rel="nofollow noopener">github.com</a> sometimes works, but mostly fails.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/4ab540cf4d064a839f33926b8601d15453942ef4.png" data-download-href="/uploads/short-url/aETFhOpCplkoXQ9n3WXdZu8kboo.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/9629574/181087695-adf6dadf-f384-4fa7-a6c0-7cfab9a2b90f.png" alt="image" data-base62-sha1="aETFhOpCplkoXQ9n3WXdZu8kboo" width="473" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/4/4ab540cf4d064a839f33926b8601d15453942ef4_2_473x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/4/4ab540cf4d064a839f33926b8601d15453942ef4_2_709x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/4/4ab540cf4d064a839f33926b8601d15453942ef4.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/4ab540cf4d064a839f33926b8601d15453942ef4_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">899×949 59.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/3/3d86eda2210e3268ed49882f2011c31fb350511f.png" data-download-href="/uploads/short-url/8MicU8adYwmQjNkF2z13FlpNh71.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/9629574/181087696-37ca2160-b40c-4f2d-92bd-0d193d40675e.png" alt="image" data-base62-sha1="8MicU8adYwmQjNkF2z13FlpNh71" width="550" height="499" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/3d86eda2210e3268ed49882f2011c31fb350511f_2_550x499.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/3d86eda2210e3268ed49882f2011c31fb350511f_2_825x748.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/3/3d86eda2210e3268ed49882f2011c31fb350511f.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/3d86eda2210e3268ed49882f2011c31fb350511f_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1043×947 49.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/e5838de5bbe2c8f97e782131faf15781d6244fb5.png" data-download-href="/uploads/short-url/wKnc3CWlBzy4xAtTLpQWvlw9IOh.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/9629574/181087698-907a74f2-7159-454b-94c0-eac32658f60d.png" alt="image" data-base62-sha1="wKnc3CWlBzy4xAtTLpQWvlw9IOh" width="690" height="146" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/e5838de5bbe2c8f97e782131faf15781d6244fb5_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">879×186 10.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>So what exactly is going on here?</p>

**Comment by yaselc** [comment]
<p>In my case, the answer has been most of the time in the content of the <code>PackageReleaseNotes</code> tag, by using characters like <code>(</code>, <code>)</code>, <code>'</code> has caused the publication of the package failed.</p>

**Comment by PieterjanDeClippel** [comment]
<p>Thanks for the tip. However I removed the <code>PackageReleaseNotes</code> tag now and it’s still only working once in a whiile.</p>
<p>I have enabled diagnostic output and can see that the workflow actually pulls the github.token from the workflow, so authentication should also not be an issue.</p>
<p>Successful push:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/e4c76632a019cc29de8781265b96d4b2c3f57680.png" data-download-href="/uploads/short-url/wDS4APZ2PzR9cE4RCmy3bucnoU8.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/9629574/181087701-6162e8f4-aeca-4cda-97ea-812e7d6951e7.png" alt="image" data-base62-sha1="wDS4APZ2PzR9cE4RCmy3bucnoU8" width="690" height="457" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/e4c76632a019cc29de8781265b96d4b2c3f57680_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">835×554 31.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Failed push:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/c/cf9426ffba298822427858997ef793f43e47158a.png" data-download-href="/uploads/short-url/tCkfA1p1z8ckMbaVeieEEyWJ95U.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/9629574/181087702-8d7bafc1-d24a-4934-84ef-cf9ff62ecfe9.png" alt="image" data-base62-sha1="tCkfA1p1z8ckMbaVeieEEyWJ95U" width="689" height="316" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/c/cf9426ffba298822427858997ef793f43e47158a_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1195×549 21.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div><br>
<div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/6/6477897674a8ccb4bf21a0868d19b86daf923fa6.png" data-download-href="/uploads/short-url/ekLQW2uh6LM4D5msRwx332TghD0.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/9629574/181087704-53068aee-1ac1-4ffa-8566-c4c082fe88f7.png" alt="image" data-base62-sha1="ekLQW2uh6LM4D5msRwx332TghD0" width="690" height="272" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/6/6477897674a8ccb4bf21a0868d19b86daf923fa6_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1145×453 21.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>I also tried swapping the Github/Nuget push but the workflow still fails on PushGithub most of the time.</p>

**Comment by jcansdale** [comment, marked as answer]
<p>Hi <a class="mention" href="/pieterjandeclippel">@PieterjanDeClippel</a>,</p>
<p>This appears to be an issue with <code>Transfer-Encoding: chunked</code> HTTP requests. I tried <a href="https://github.com/jcansdale-test/NuGet.Client/commit/d22bdfb38746715a0b8d451709d4374eb4658402" rel="nofollow noopener">tweaking</a> the NuGet source to not use chunked encoding and the push works every time!<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/jcansdale-test/NuGet.Client/actions/runs/237460162/workflow" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/11719160/181087707-3c2010c8-4a9d-4852-bdc3-b0c104ec1c0a.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/jcansdale-test/NuGet.Client/actions/runs/237460162/workflow" target="_blank" rel="nofollow noopener">jcansdale-test/NuGet.Client</a></h3>

<p>Client Tools for NuGet - including Visual Studio extensions, command line tools, and msbuild support. (Open issues on https://github.com/nuget/home/issues) - jcansdale-test/NuGet.Client</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Bizarrely it does work with <code>nuget.exe</code> on the .NET framework, which also claims to use <code>Transfer-Encoding: chunked</code>. My hunch is that this sends the payload as a single chunk.</p>
<p>Could you try modifying your workflow to do this:</p>
<pre><code class="lang-auto">    - name: PushGithub
      run: nuget.exe push *.nupkg --no-symbols --skip-duplicate
      env:
        NUGET_AUTH_TOKEN: ${{ github.token }}
</code></pre>
<p>You <em>should</em> find this will work consistently.</p>
<p>Please let me know how you get on!</p>

**Comment by PieterjanDeClippel** [comment]
<p>Thanks for the response.</p>
<p>I’ve tried using <code>nuget.exe</code> instead. It has run a first time and made a successful push. I’ll need to try some more pushes to draw a definite conclusion.</p>
<p>I have to note however that I’ve got a similar setup with a pure .NET Standard project, and this project has never given a problem building and pushing to nuget and GPR. I suspect that the issue is introduced when publishing a multi-target nuget package (I have no clue what could possibly cause this).</p>
<p>I’ll keep you up-to-date if using nuget.exe works consistently.</p>
<p>(6/6 successful pushes)</p>

**Comment by jcansdale** [comment]
<p>The chance of a <code>dotnet nuget push</code> failing seems to increase with the size of the package (but it’s still somewhat random).</p>
<p>Here is an example that shows pushes of various sized packages:<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/jcansdale-test/nuget-push-test/actions/runs/244671907/workflow" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/11719160/181087708-1d7999b1-705b-42cc-820d-de22274921b2.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/jcansdale-test/nuget-push-test/actions/runs/244671907/workflow" target="_blank" rel="nofollow noopener">jcansdale-test/nuget-push-test</a></h3>

<p>Contribute to jcansdale-test/nuget-push-test development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>By the time we get to a 2049 kB package, the push is almost certain to fail!</p>

**Comment by PieterjanDeClippel** [comment]
<p>Interesting finding. I never had a problem pushing a .NET Standard nuget package to <a href="http://github.com/nuget.org" rel="nofollow noopener">github.com/nuget.org</a>. The problem started occuring by creating a multi-target nuget package. This implies that for each target, there’s a DLL in the package and increases the size of the package drastically.</p>
<p>But how come the package can successfully be pushed to <a href="http://nuget.org" rel="nofollow noopener">nuget.org</a> all the time, while it fails for <a href="http://github.com" rel="nofollow noopener">github.com</a>, using the same command?</p>
<p>Anyhow, using nuget.exe seems to work reliably. Thanks for the help.</p>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="PieterjanDeClippel" data-post="8" data-topic="131032">
<div class="title">
<div class="quote-controls"></div>
 PieterjanDeClippel:</div>
<blockquote>
<p>But how come the package can successfully be pushed to <a href="http://nuget.org" rel="nofollow noopener">nuget.org</a> all the time, while it fails for <a href="http://github.com" rel="nofollow noopener">github.com</a>, using the same command?</p>
</blockquote>
</aside>
<p>It’s probably something to do with GitHub’s load balancer and they way chunked encoding works. The team will hopefully be digging into it soon!</p>

**Comment by ckapatch** [comment]
<p>This definitely has something to do with the size of the package. I have smaller packages that upload without issue but bigger ones fail with the above error message. It’s really really inconvenient and frustrating.</p>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="ckapatch" data-post="10" data-topic="131032">
<div class="title">
<div class="quote-controls"></div>
 ckapatch:</div>
<blockquote>
<p>It’s really really inconvenient and frustrating.</p>
</blockquote>
</aside>
<p>I’ve written a GitHub Action based on <a href="https://github.com/jcansdale/gpr" rel="noopener nofollow ugc">https://github.com/jcansdale/gpr</a> that you should find works consistently.</p>
<p>You can use it like this:</p>
<pre><code class="lang-auto">    - uses: jcansdale/gpr@v0.1.257
      with:
        command: push
        path: **/*.nupkg
</code></pre>
<p>Please let me know if you try it!</p>

**Comment by ckapatch** [comment]
<p>Great! Thanks. I’ll check this out.</p>

**Comment by yrien30** [comment]
<p>Hi, thanks for this Github Action. Just note that if you build a dotnet package whit --include-symbols then it will upload each package two times and complain that its allready uploaded second time it gets uploaded.</p>

**Comment by yaselc** [comment]
<p>I replaced <code>dotnet nuget</code> by <a class="mention" href="/jcansdale">@jcansdale</a> tool for working with the GitHub Package Registry and it works like a charm:</p>
<pre><code class="lang-yaml">      - name: Install tool for working with the GPR
        run: dotnet tool install -g gpr

      - name: Publish Package
        run: |
          gpr push PATH-TO-PKG-FILE.nupkg
</code></pre>

**Comment by mtttcgcg** [comment]
<p>Hello, we are seeing this issue with Nuget packages &lt; 50 KB in size. It would be nice to avoid having to use the gpr push workaround. Do we know what the root cause for this failure is when using dotnet nuget push?  There are no commas in my package description.</p>

**Comment by jcansdale** [comment]
<p>Hi <a class="mention" href="/mtttcgcg">@mtttcgcg</a>,</p>
<aside class="quote no-group" data-username="mtttcgcg" data-post="15" data-topic="131032">
<div class="title">
<div class="quote-controls"></div>
 mtttcgcg:</div>
<blockquote>
<p>Hello, we are seeing this issue with Nuget packages &lt; 50 KB in size. It would be nice to avoid having to use the gpr push workaround. Do we know what the root cause for this failure is when using dotnet nuget push?</p>
</blockquote>
</aside>
<p>If you use the <code>dotnet nuget push --api-key &lt;PAT&gt;</code> option, you should find pushes work consistently. We don’t know what the root cause for this failure was, but it appears to be somehow authentication related. You can continue using <code>nuget.config</code> for downloading packages. This failure was only an issue for pushes.</p>
<p>I hope that helps!</p>
